internal/util/ssa/cache.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssa
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"sigs.k8s.io/cluster-api/internal/util/hash"
+)
+
+const (
+	// ttl is the duration for which we keep the keys in the cache.
+	ttl = 10 * time.Minute
+
+	// expirationInterval is the interval in which we will remove expired keys
+	// from the cache.
+	expirationInterval = 10 * time.Hour
+)
+
+// Cache caches SSA request results.
+// Specifically we only use it to cache that a certain request
+// doesn't have to be repeated anymore because there was no diff.
+type Cache interface {
+	// Add adds the given key to the Cache.
+	// Note: keys expire after the ttl.
+	Add(key string)
+
+	// Has checks if the given key (still) exists in the Cache.
+	// Note: keys expire after the ttl.
+	Has(key string) bool
+}
+
+// NewCache creates a new cache.
+func NewCache() Cache {
+	r := &ssaCache{
+		Store: cache.NewTTLStore(func(obj interface{}) (string, error) {
+			// We only add strings to the cache, so it's safe to cast to string.
+			return obj.(string), nil
+		}, ttl),
+	}
+	go func() {
+		for {
+			// Call list to clear the cache of expired items.
+			r.List()
+
+			time.Sleep(expirationInterval)
+		}
+	}()
+	return r
+}
+
+type ssaCache struct {
```

Comment:
I'm fine with changing the struct name if someone has a better idea. I could not use "cache" as that would conflict with the imported package and it feels strange to use an import alias for cache because of that.